include/lock/optimistic_lock.hpp

```diff
@@ -0,0 +1,233 @@
+   * corrupt an internal lock status.
+   */
+  void
+  DowngradeToSIX()
+  {
+    lock_.store(kSIXLock, std::memory_order_release);
+  }
+
+  /**
+   * @brief Release an exclusive lock.
+   *
+   */
+  void
+  UnlockX()
+  {
+    auto expected = lock_.load(std::memory_order_relaxed);
+    auto desired = expected + kXLock;
+
+    while (!lock_.compare_exchange_weak(expected, desired, std::memory_order_acquire,
+                                        std::memory_order_relaxed)) {
+      desired = expected + kXLock;
+      SPINLOCK_HINT
+    }
+  }
+
+  /**
+   * @brief Get a shared lock with an intent-exclusive lock.
+   *
+   * NOTE: this function does not give up acquiring locks and continues spinlock.
+   */
+  void
+  LockSIX()
+  {
+    while (true) {
+      auto expected = lock_.load(std::memory_order_relaxed) & kSIXLockMask;
+      auto desired = expected | kSIXLock;
+      for (size_t i = 0; i < kRetryNum; ++i) {
+        const auto cas_success = lock_.compare_exchange_weak(
+            expected, desired, std::memory_order_acquire, std::memory_order_release);
+        if (cas_success) return;
+
+        expected &= kSIXLockMask;
+        desired = expected | kSIXLock;
+        SPINLOCK_HINT
+      }
+
+      std::this_thread::sleep_for(kShortSleep);
+    }
+  }
+
+  /**
+   * @brief Upgrade an SIX lock to an X lock.
+   *
+   * NOTE: if a thread that does not have a shared lock with an intent-exclusive lock
+   * calls this function, it will corrupt an internal lock status.
+   */
+  void
+  UpgradeToX()
+  {
+    while (true) {
+      auto expected = kSIXLock;
+      for (size_t i = 0; i < kRetryNum; ++i) {
+        const auto cas_success = lock_.compare_exchange_weak(
+            expected, kXLock, std::memory_order_acquire, std::memory_order_relaxed);
+        if (cas_success) return;
+
+        expected = kSIXLock;
+        SPINLOCK_HINT
+      }
+
+      std::this_thread::sleep_for(kShortSleep);
+    }
+  }
+
+  /**
+   * @brief Release a shared lock with an intent-exclusive lock.
+   *
+   */
+  void
+  UnlockSIX()
+  {
+    auto expected = lock_.load(std::memory_order_relaxed);
+    auto desired = expected - kSIXLock;
+    while (!lock_.compare_exchange_weak(expected, desired, std::memory_order_relaxed)) {
+      desired = expected - kSIXLock;
+      SPINLOCK_HINT
+    }
+  }
+
+ private:
+  /*####################################################################################
+   * Internal member variables
+   *##################################################################################*/
+
+  /// a lock status for no locks.
+  static constexpr uint64_t kNoLocks = 0b000UL;
+
+  /// a lock status for exclusive locks.
+  static constexpr uint64_t kXLock = 0b010UL;
+
+  /// a lock status for shared locks with intent-exclusive locks.
+  static constexpr uint64_t kSIXLock = 0b001UL;
+
+  /// a bit mask for removing SIX/X-lock flags.
+  static constexpr uint64_t kSIXLockMask = ~0b011UL;
```

Comment:
どっちでもいいですが，XLockMaskにしましょうか．

```suggestion
  static constexpr uint64_t kXLockMask = ~0b011UL;
```